src/Common/src/Interop/Unix/libssl/SecuritySafeHandles.cs

```diff
@@ -207,13 +214,49 @@ protected override bool ReleaseHandle()
     }
 
 #if DEBUG
-    internal sealed class SafeDeleteContext : DebugSafeHandle
+    internal abstract class SafeDeleteContext : DebugSafeHandle
     {
 #else
-    internal sealed class SafeDeleteContext : SafeHandle
+    internal abstract class SafeDeleteContext : SafeHandle
     {
 #endif
-        private readonly SafeFreeCredentials _credential;
+        private SafeFreeCredentials _credential;
+
+        protected SafeDeleteContext(SafeFreeCredentials credential)
+            : base(IntPtr.Zero, true)
+        {
+            Debug.Assert((null != credential), "Invalid credential passed to SafeDeleteContext");
+
+            // When a credential handle is first associated with the context we keep credential
+            // ref count bumped up to ensure ordered finalization. The credential properties
+            // are used in the SSL/NEGO data structures and should survive the lifetime of
+            // the SSL/NEGO context
+            bool ignore = false;
+            _credential = credential;
+            _credential.DangerousAddRef(ref ignore);
+        }
+
+        public override bool IsInvalid
+        {
+            get { return (null == _credential); }
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            Debug.Assert((null != _credential), "Null credential in SafeDeleteContext");
+            _credential.DangerousRelease();
+            _credential = null;
+            return true;
+        }
+
+        public override string ToString()
+        {
+            return IsInvalid ? String.Empty : handle.ToString();
```

Comment:
Yes.. this debug info was present in the Windows SafeHandle so we retained it since the platform-agnostic code uses it in GlobalLog.Print statements.

@shrutigarg I guess it can be changed to handle.ToString